src/core_plugins/kibana/public/visualize/listing/visualize_landing_page_table.js

```diff
@@ -162,43 +118,30 @@ export class VisualizeLandingPageTable extends React.Component {
     return this.getColumnSortOrder(TYPE_COLUMN_ID);
   }
 
-  getAreAllItemsSelected() {
-    const { selectedIds, items, currentPage } = this.state;
-    const pageOfItems = this.pager.getItemsOnPage(items, currentPage);
-    const pageOfItemIds = pageOfItems.map(item => item.id);
-    return areAllItemsSelected(selectedIds, pageOfItemIds);
+  areAllItemsSelected() {
+    return areAllItemsSelected(this.state, this.pager);
   }
 
-  getVisualizeColumns() {
+  renderVisualizeColumns() {
     const { selectedIds } = this.state;
     return [
-      getCheckBoxColumn(this.getAreAllItemsSelected(), selectedIds, this.onToggleItem, this.onToggleAll),
-      getTitleColumn(this.getEditUrlForItem, this.getTitleSortOrder(), this.onSortByTitle),
-      getTypeColumn(this.onSortByType, this.getTypeSortOrder()),
+      getCheckBoxColumn(this.areAllItemsSelected(), selectedIds, this.onToggleItem, this.onToggleAll),
+      getTitleColumn(this.getEditUrlForItem, this.getTitleSortOrder(), this.onSort.bind(null, TITLE_COLUMN_ID)),
+      getTypeColumn(this.onSort.bind(null, TYPE_COLUMN_ID), this.getTypeSortOrder()),
     ];
   }
 
-  onDelete = () => {
-    const { deleteItems } = this.props;
-    const { selectedIds, filter } = this.state;
-    deleteItems(selectedIds).then((didDelete) => {
-      if (didDelete) {
-        this.onFilter(filter);
-      }
-    });
-  };
-
-  getActionButtons() {
+  renderActionButtons() {
     return this.state.selectedIds.length > 0
       ? <DeleteButton onClick={this.onDelete} />
       : <CreateButtonLink href={'#' + VisualizeConstants.WIZARD_STEP_1_PAGE_PATH} />;
   }
 
-  getTableContents() {
+  renderTableContents() {
```

Comment:
I like the row-centric approach, because it appears to be more powerful (also applies to other representations like list-of-cards). Creating the whole table body in one go though feels very restrictive. It would make stuff like infinite loading and virtual scrolling a lot more inefficient/difficult to implement. How about providing "blueprints" for the header and rows to the table, which it "instantiates" using [`React.cloneElement`](https://facebook.github.io/react/docs/react-api.html#cloneelement). This way jsx provides a declarative way to configure the table. Different representations for different rows can then be delegated to row components that choose a particular way of rendering depending on the document:

(just a quick draft, expect typos)

```javascript
class DashboardLandingPageTable {
  render() {
    const headerRow = (
      <ItemTableHeaderRow>
        <ItemTableHeader
          label="Name"
          onChangeSort={this.actions.sortBy("name")}
          // ...
        />
        <ItemTableHeader
          label="Created At"
          onChangeSort={this.actions.sortBy("creation_date")}
          // ...
        />
        // ...
      </ItemTableHeaderRow>
    );

    const bodyRow = (
      <ItemTableBodyRow>
        <ItemTableStringCell
          key="name"
        />
        <ItemTableDateCell
          key="creation_date"
          format="YYYY-mm-dd"
        />
      </ItemTableBodyRow>
    );

    return (
      <ItemTable
        bodyRow={bodyRow}
        headerRow={headerRow}
        items={items}
      />
    );
  }
}

function ItemTable({bodyRow, headerRow, items}) {
  return (
    <table class="kuiItemTable">
      <thead>
        { React.cloneElement(headerRow) }
      </thead>
      <tbody>
        { items.map((item) => React.cloneElement(bodyRow, { item })) }
      </tbody>
    </table>
  );
}

function ItemTableBodyRow({ children, item }) {

  if (item.confidential) {
    return (
      <tr class="">
        <td>You do not need to know</td>
      </tr>
    );
  } else {
    return (
      <tr class="tableRow--highlight">
        { React.Children.map(children, (child) => React.cloneElement(child, { item })) }
      </tr>
    );
  }
}

function ItemTableDateCell({ format, item, key }) {
  return (
    <td>
      { formatDate(item[key], format) }
    </td>
  );
}
```